### PR TITLE
scalatra metrics and health checks

### PR DESCRIPTION
**Pull request by sokrahta**
What's new
----
+ add scalatra-metrics
+ rev scalatra to 2.4.0.M3
+ rev scala to 2.10.5
+ mount metrics endpoints
+ instrument metrics on /suggest and /sample endpoints
+ add timers around endpoints with external calls
+ add servlet top level error and notFound counters
+ add basic health checks

Results
----
Start from [metrics-admin](http://localhost:8042/metrics-admin)
Click link to [Healthcheck](http://localhost:8042/metrics-admin/healthcheck?pretty=true)
Each of these healthchecks runs a code block periodically and registers the response.
```
{
  "com.socrata.spandex.http.SpandexServlet.alive" : {
    "healthy" : true
  },
  "com.socrata.spandex.http.SpandexServlet.esClusterHealth" : {
    "healthy" : true,
    "message" : "{\n  \"cluster_name\" : \"spandex\",\n  \"status\" : \"yellow\",\n  \"timed_out\" : false,\n  \"number_of_nodes\" : 1,\n  \"number_of_data_nodes\" : 1,\n  \"active_primary_shards\" : 5,\n  \"active_shards\" : 5,\n  \"relocating_shards\" : 0,\n  \"initializing_shards\" : 0,\n  \"unassigned_shards\" : 5\n}"
  },
  "com.socrata.spandex.http.SpandexServlet.version" : {
    "healthy" : true,
    "message" : "{\"name\":\"spandex-http\",\"buildTime\":\"1429031694175\",\"scalaVersion\":\"2.10.5\",\"revision\":\"v0.7.0-1-g8b5a8f2648-dirty\",\"version\":\"0.7.1-SNAPSHOT\",\"sbtVersion\":\"0.13.7\"}"
  }
}
```
Click link to [Metrics](http://localhost:8042/metrics-admin/metrics?pretty=true)
Instrumented filter adds a bunch of metrics by default. In addition we've added the servlet top level error/notFound counters and suggest/sample timers.
```
{
  "version" : "3.0.0",
  "gauges" : { },
  "counters" : {
    "com.codahale.metrics.servlet.InstrumentedFilter.activeRequests" : {
      "count" : 0
    },
    "com.socrata.spandex.http.SpandexServlet.error" : {
      "count" : 1
    },
    "com.socrata.spandex.http.SpandexServlet.notFound" : {
      "count" : 2
    }
  },
  "histograms" : { },
  "meters" : {
    "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.badRequest" : {
      "count" : 0,
      "m15_rate" : 0.0,
      "m1_rate" : 0.0,
      "m5_rate" : 0.0,
      "mean_rate" : 0.0,
      "units" : "events/second"
    },
    "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.created" : {
      "count" : 0,
      "m15_rate" : 0.0,
      "m1_rate" : 0.0,
      "m5_rate" : 0.0,
      "mean_rate" : 0.0,
      "units" : "events/second"
    },
    "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.noContent" : {
      "count" : 0,
      "m15_rate" : 0.0,
      "m1_rate" : 0.0,
      "m5_rate" : 0.0,
      "mean_rate" : 0.0,
      "units" : "events/second"
    },
    "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.notFound" : {
      "count" : 26,
      "m15_rate" : 0.025275284430951516,
      "m1_rate" : 0.07543758787934625,
      "m5_rate" : 0.05833880594928567,
      "mean_rate" : 0.05190196333313444,
      "units" : "events/second"
    },
    "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.ok" : {
      "count" : 0,
      "m15_rate" : 0.0,
      "m1_rate" : 0.0,
      "m5_rate" : 0.0,
      "mean_rate" : 0.0,
      "units" : "events/second"
    },
    "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.other" : {
      "count" : 0,
      "m15_rate" : 0.0,
      "m1_rate" : 0.0,
      "m5_rate" : 0.0,
      "mean_rate" : 0.0,
      "units" : "events/second"
    },
    "com.codahale.metrics.servlet.InstrumentedFilter.responseCodes.serverError" : {
      "count" : 1,
      "m15_rate" : 0.0011080303990206543,
      "m1_rate" : 0.015991117074135343,
      "m5_rate" : 0.0033057092356765017,
      "mean_rate" : 0.0019962265133111473,
      "units" : "events/second"
    }
  },
  "timers" : {
    "com.codahale.metrics.servlet.InstrumentedFilter.requests" : {
      "count" : 27,
      "max" : 0.091007826,
      "mean" : 0.013766574076166119,
      "min" : 0.0025895970000000003,
      "p50" : 0.008041474,
      "p75" : 0.019021785000000003,
      "p95" : 0.037698623,
      "p98" : 0.037698623,
      "p99" : 0.091007826,
      "p999" : 0.091007826,
      "stddev" : 0.015193621897202273,
      "m15_rate" : 0.02638331482997217,
      "m1_rate" : 0.09142870495348158,
      "m5_rate" : 0.06164451518496217,
      "mean_rate" : 0.0538982304669007,
      "duration_units" : "seconds",
      "rate_units" : "calls/second"
    },
    "com.socrata.spandex.http.SpandexServlet.sample" : {
      "count" : 7,
      "max" : 3.265E-6,
      "mean" : 1.5194978596190627E-6,
      "min" : 1.1320000000000001E-6,
      "p50" : 1.3270000000000002E-6,
      "p75" : 1.591E-6,
      "p95" : 3.265E-6,
      "p98" : 3.265E-6,
      "p99" : 3.265E-6,
      "p999" : 3.265E-6,
      "stddev" : 5.686172421512064E-7,
      "m15_rate" : 1.0738071801772435,
      "m1_rate" : 0.22665072340507444,
      "m5_rate" : 0.8598375726885472,
      "mean_rate" : 0.06449553569033775,
      "duration_units" : "seconds",
      "rate_units" : "calls/second"
    },
    "com.socrata.spandex.http.SpandexServlet.suggestSample" : {
      "count" : 9,
      "max" : 2.7077700000000004E-4,
      "mean" : 3.1884117003082016E-5,
      "min" : 1.0080000000000001E-6,
      "p50" : 1.4900000000000001E-6,
      "p75" : 1.709E-6,
      "p95" : 2.7077700000000004E-4,
      "p98" : 2.7077700000000004E-4,
      "p99" : 2.7077700000000004E-4,
      "p999" : 2.7077700000000004E-4,
      "stddev" : 8.380285593966399E-5,
      "m15_rate" : 1.5406911421420764,
      "m1_rate" : 0.1745495421559294,
      "m5_rate" : 1.1287603534915014,
      "mean_rate" : 0.06155541066442672,
      "duration_units" : "seconds",
      "rate_units" : "calls/second"
    },
    "com.socrata.spandex.http.SpandexServlet.suggestText" : {
      "count" : 9,
      "max" : 1.8940000000000002E-6,
      "mean" : 1.2915436307301368E-6,
      "min" : 1.0420000000000001E-6,
      "p50" : 1.175E-6,
      "p75" : 1.3290000000000001E-6,
      "p95" : 1.8940000000000002E-6,
      "p98" : 1.8940000000000002E-6,
      "p99" : 1.8940000000000002E-6,
      "p999" : 1.8940000000000002E-6,
      "stddev" : 2.4281814776159496E-7,
      "m15_rate" : 1.566584506500102,
      "m1_rate" : 0.2241260485994217,
      "m5_rate" : 1.1866331343607992,
      "mean_rate" : 0.06730715128914287,
      "duration_units" : "seconds",
      "rate_units" : "calls/second"
    }
  }
}
```

Also: [Ping](http://localhost:8042/metrics-admin/ping)
And more: [Threads](http://localhost:8042/metrics-admin/threads)

[Review on Reviewable](https://reviewable.io/reviews/socrata/spandex/45)

### CODE REVIEW

**Comment by sokrahta** [comment]
[Performance smoketest results](https://docs.google.com/document/d/1fhnqdXynuOe2qFNKjdkBTI5y6Hy7eUClAaz7FyPG4SY/edit)

File: spandex-http/src/main/scala/com/socrata/spandex/http/SpandexStack.scala [hunk 1]
```diff
@@ -40,6 +46,7 @@ trait SpandexStack extends ScalatraServlet with ScalateSupport with Logging {
 
   error {
     case e: Exception =>
+      counter("error") += 1
```
Comment (velvia) [inline spandex-http/src/main/scala/com/socrata/spandex/http/SpandexStack.scala]: You might want to put a timer here as well.  That way you can compare rate of errors vs rate of successes.

**Comment by velvia** [comment]
Scalatra-metrics looks great.  We should really standardize on a set of admin routes though (beyond the ubiquitous /version)

File: spandex-http/src/main/scala/com/socrata/spandex/http/SpandexStack.scala (hunk 1 above)
Comment (sokrahta) [inline spandex-http/src/main/scala/com/socrata/spandex/http/SpandexStack.scala]: I could replace counter with timer.

**Pull request merged**